torch/_inductor/spmd_prefetch.py

```diff
@@ -0,0 +1,128 @@
+# mypy: allow-untyped-defs
+# pyre-strict
+import typing
+from typing import (
+    Any,
+    Counter,
+    DefaultDict,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+from . import config, ir, comms, scheduler
+from collections import OrderedDict
+
+import torch
+
+from .dependencies import WeakDep
+from .utils import is_collective, is_wait, tuple_sorted
+
+
+    
+def reorder_all_gather(
+    snodes: List["scheduler.BaseSchedulerNode"],
+    all_gather_order: Optional[str] = "before",
+) -> List["scheduler.BaseSchedulerNode"]:
+    '''
+    Reorder All Gather and Wait in the forward/backward pass;
+    1. all_gather_order set to before: all_gather_i is reordered before wait_i-1
+    2. all_gather_order set to after: all_gather_i is reordered after wait_i-1
+    '''
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
+    all_gather_list: List["scheduler.BaseSchedulerNode"] = []
+    node_type_list: List[int] = []
+
+    inverse_users, node_users = comms.compute_node_users(snodes)
+
+    for node in reversed(snodes):
+        node_type_list.append(get_node_type(node))
+
+    for idx, node in enumerate(reversed(snodes)):
```

Comment:
The two reordering algorithms look non-straightforward, especially with the integer node types.
After changing node types to use Enum, would you please add comments to code to explain how you achieve the desired order? Good code readability makes it easy to understand and debug.